docs/design/operator.md

```diff
@@ -0,0 +1,383 @@
+deployment, configuration, and management of these drivers using new Kubernetes
+APIs defined as a set of Custom Resource Definitions (CRDs).
+
+## System Overview
+
+The Ceph-CSI Operator is a Kubernetes operator that offers a Kubernetes native
+management interfaces for Ceph CSI drivers. It provides a set of new Kubernetes
+APIs (CRDs) for governing deployment and management of Ceph-CSI drivers, as
+well as connection details to Ceph clusters. The operator ensures that the
+desired state specified in the CRDs is maintained in the cluster.
+
+## Assumptions and Dependencies
+
+Kubernetes cluster is available and operational. Ceph cluster is set up and
+accessible from the Kubernetes cluster. Users are familiar with Kubernetes and
+Ceph concepts. CRD and operator versions are compatible with the Kubernetes
+version in use.
+
+## Goals and Guidelines
+
+- Automate the deployment and lifecycle for Ceph-CSI drivers.
+- Provide a consistent and flexible configuration mechanism using CRDs.
+- Enable namespace-scoped configurations for different Ceph-CSI drivers.
+- Ensure high availability and scalability of the operator.
+- Support multiple deployment modes for the operator:
+  - Single-namespace: the operator manages a single Ceph-CSI driver in the
+    current namespace where operator is deployed.
+  - Multi-namespace: the operator manages Ceph-CSI drivers in user-specified
+    namespaces.
+  - All-namespaces: the operator manages Ceph-CSI drivers in all namespaces.
+
+## Development Methods
+
+Use operator framework which is higly used in operator developed and also by
+following Kubernetes best practices for operator development.
+
+## Architecture
+
+```mermaid
+graph TD
+    A[CRD Changes] --> B[Operator]
+    B -->  D[Configure CephFS CSI Driver]
+    B -->  E[Configure RBD CSI Driver]
+    B -->  F[Configure NFS CSI Driver]
+
+```
+
+In this diagram:
+
+- **CRD Changes**: Represents changes made to Custom Resource Definitions,
+  which trigger actions in the operator.
+- **Operator**: Listens for changes in CRDs and initiates the installation and
+  configuration of CSI drivers.
+- **Configure CephFS, NFS, RBD**: Actions performed by the operator to install
+  and configure the respective CSI drivers based on the CRD changes.
+
+## CRDs for ceph-csi-operator
+
+### CephCSIOperatorConfig CRD
+
+Manages operator-level configurations and offers a place to overwrite settings
+for CSI drivers. This CRD is a namespace-scoped CRD and a single CR named
+ceph-csi-operator-config can be created by the admin in the namespace
+where operator is running. The configurations are catagorized into 2 different
+parts
+
+- Operator configurations
+- Common configurations across all CSI driver
+
+```yaml
+---
+kind: CephCSIOperatorConfig
+apiVersion: csi.ceph.io/v1
+metadata:
+  name: csioperatorconfig
+  namespace: operator-namespace
+spec:
+  logLevel: 1
+  driverSpecDefaults:
+    logging:
```

Comment:
Does this logging setting go to new CRDS, (all the csi containers)?

I think we can have a separate section
```
logCollector:
    enabled: true
    periodicity: daily # one of: hourly, daily, weekly, monthly
    maxLogSize: 500M # SUFFIX may be 'M' or 'G'. Must be at least 1M.
```
We should keep it similar way that we have for ceph daemons

Reply:
This is a separate section for logging but we don't have any plan to have common keys that represent ceph as these are for sidecar and cephcsi logging we will try to keep the configuration fields similar to Linux/kubernetes log rotation. we can  change the fields later on once we have a solid design that you are working on related to the configurations.